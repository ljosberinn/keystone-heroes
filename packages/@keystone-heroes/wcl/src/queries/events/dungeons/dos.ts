import type { AllTrackedEventTypes, ApplyDebuffEvent } from "../types";
import { createIsSpecificEvent } from "../utils";

export const DOS_URN = 228_626;

/**
 * @see https://www.warcraftlogs.com/reports/hFj3wLzrapC4KvZk#fight=2&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24type%20%3D%20%22applydebuff%22%20and%20target.type%20%3D%20%22NPC%22%20and%20ability.id%20%3D%20228626
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "hFj3wLzrapC4KvZk") {
 *       fights(fightIDs: [2]) {
 *         startTime
 *         endTime
 *       }
 *        events(startTime: 34007, endTime: 1759795, filterExpression: "type = \"applydebuff\" and target.type = \"npc\" and ability.id = 228626") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "applydebuff" and target.type = "npc" and ability.id = ${DOS_URN}`,
];

const isDosUrnEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: DOS_URN,
});

export const getDOSEvents = (
  allEvents: AllTrackedEventTypes[]
): ApplyDebuffEvent[] => {
  return allEvents.filter(isDosUrnEvent);
};
