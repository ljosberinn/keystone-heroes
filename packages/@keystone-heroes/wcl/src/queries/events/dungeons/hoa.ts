import type { AllTrackedEventTypes, CastEvent } from "../types";
import { createIsSpecificEvent } from "../utils";

export const HOA_GARGOYLE = 342_171 as const;

/**
 * @see https://www.warcraftlogs.com/reports/hFj3wLzrapC4KvZk#fight=4&type=summary&pins=2%24Off%24%23244F4B%24expression%24type%20%3D%20%22cast%22%20and%20source.type%20%3D%20%22player%22%20and%20ability.id%20%3D%20342171&view=events
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "hFj3wLzrapC4KvZk") {
 *       fights(fightIDs: [4]) {
 *         startTime
 *         endTime
 *       }
 *        events(startTime: 5987472, endTime: 7736379, filterExpression: "type = \"cast\" and source.type = \"player\" and ability.id = 342171") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "cast" and source.type = "player" and ability.id = ${HOA_GARGOYLE}`,
];

const isHoaGargoyleEvent = createIsSpecificEvent<CastEvent>({
  type: "cast",
  abilityGameID: HOA_GARGOYLE,
});

export const getHOAEvents = (
  allEvents: AllTrackedEventTypes[]
): CastEvent[] => {
  return allEvents.filter(isHoaGargoyleEvent);
};
