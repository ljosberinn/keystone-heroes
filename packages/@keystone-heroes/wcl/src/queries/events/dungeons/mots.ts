import type { ApplyDebuffEvent, AllTrackedEventTypes } from "../types";
import { createChunkByThresholdReducer, createIsSpecificEvent } from "../utils";

export const ENVELOPMENT_OF_MISTS = 323_881 as const;

/**
 * @see https://www.warcraftlogs.com/reports/aT8ZhP1HRfJANKCj#fight=18&type=summary&view=events&pins=2%24Off%24%23909049%24expression%24type%20%3D%20%22applydebuff%22%20and%20target.type%20%3D%20%22player%22%20and%20ability.id%20%3D%20323881
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "aT8ZhP1HRfJANKCj") {
 *       fights(fightIDs: [18]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 34782210, endTime: 36664869, filterExpression: "type = \"applydebuff\" and target.type = \"player\" and ability.id = 323881") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "applydebuff" and target.type = "player" and ability.id = ${ENVELOPMENT_OF_MISTS}`,
];

const isMistsMazeDebuffEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: ENVELOPMENT_OF_MISTS,
});

// failing the mists maze applies a debuff for 4 seconds
// hence its impossible to trigger it 2x within that timeframe
const debuffReducer = createChunkByThresholdReducer(4 * 1000);

export const getMOTSEvents = (
  allEvents: AllTrackedEventTypes[]
): ApplyDebuffEvent[] => {
  return (
    allEvents
      .filter(isMistsMazeDebuffEvent)
      .reduce<ApplyDebuffEvent[][]>(debuffReducer, [])
      // pick only the first event of each chunk,
      // indicating when the maze was failed
      .flatMap((chunk) => chunk[0])
  );
};
