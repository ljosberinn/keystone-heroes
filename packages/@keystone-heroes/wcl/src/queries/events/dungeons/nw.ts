import type { AllTrackedEventTypes, DamageEvent, HealEvent } from "../types";
import {
  createChunkByThresholdReducer,
  createIsSpecificEvent,
  reduceEventsByPlayer,
} from "../utils";

export const NW = {
  ORB: 328_406,
  HAMMER: 328_128,
  SPEAR: 328_351,
  KYRIAN_ORB_HEAL: 344_422,
  KYRIAN_ORB_DAMAGE: 344_421,
} as const;

/**
 * @see https://www.warcraftlogs.com/reports/Jq7KrbYV1hmTWMyw#fight=4&type=summary&view=events&pins=2%24Off%24%23909049%24expression%24type%20%3D%20%22heal%22%20and%20source.type%20%3D%20%22player%22%20and%20ability.id%20%3D344422%5E2%24Off%24%23909049%24expression%24type%20%3D%20%22damage%22%20and%20source.type%20%3D%20%22player%22%20and%20ability.id%20%3D328406%5E2%24Off%24%23a04D8A%24expression%24type%20%3D%20%22damage%22%20and%20source.type%20%3D%20%22player%22%20and%20ability.id%20%3D328128%5E2%24Off%24%23DF5353%24expression%24type%20%3D%20%22damage%22%20and%20source.type%20%3D%20%22player%22%20and%20ability.id%20%3D328351%5E2%24Off%24rgb(78%25,%2061%25,%2043%25)%24expression%24type%20%3D%20%22damage%22%20and%20source.type%20%3D%20%22player%22%20and%20ability.id%20%3D344421
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "Jq7KrbYV1hmTWMyw") {
 *       fights(fightIDs: [4]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 1764363, endTime: 3716117, filterExpression: "(type = \"heal\" and source.type = \"player\" and ability.id = 344422) or (type = \"damage\" and source.type = \"player\" and ability.id in (328406, 328128, 328351, 344421))") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "heal" and source.type = "player" and ability.id = ${NW.KYRIAN_ORB_HEAL}`,
  `type = "damage" and source.type = "player" and ability.id in (${Object.values(
    NW
  )
    .filter((value) => value !== NW.KYRIAN_ORB_HEAL)
    .join(", ")})`,
];

const isNwSpearEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: NW.SPEAR,
});

const isNwHammerEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: NW.HAMMER,
});

const isNwOrbEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: NW.ORB,
});

const isNwKyrianOrbDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: NW.KYRIAN_ORB_DAMAGE,
});

const isNwKyrianOrbHealEvent = createIsSpecificEvent<HealEvent>({
  type: "heal",
  abilityGameID: NW.KYRIAN_ORB_HEAL,
});

// NW Spear applies a bleed for 16 seconds
// each usage is hopefully thus at least 16s apart of each other
// may have to adjust later for multi-spearing...
const nwSpearReducer = createChunkByThresholdReducer(16 * 1000);

// NW Orb pulses every 1s for 8s
// each usage is hopefully thus at least 8s apart of each other
// may have to adjust later for multi-orbing...
const nwOrbReducer = createChunkByThresholdReducer(8 * 1000);

export const getNWEvents = (
  allEvents: AllTrackedEventTypes[]
): (DamageEvent | HealEvent)[] => {
  return [
    ...allEvents
      .filter(isNwSpearEvent)
      .reduce<DamageEvent[][]>(nwSpearReducer, [])
      .flatMap((chunk) => reduceEventsByPlayer(chunk, "sourceID")),
    ...allEvents
      .filter(isNwOrbEvent)
      .reduce<DamageEvent[][]>(nwOrbReducer, [])
      .flatMap((chunk) => reduceEventsByPlayer(chunk, "sourceID")),
    ...allEvents.filter(isNwHammerEvent),
    ...reduceEventsByPlayer(
      allEvents.filter(isNwKyrianOrbDamageEvent),
      "sourceID"
    ),
    ...reduceEventsByPlayer(
      allEvents.filter(isNwKyrianOrbHealEvent),
      "sourceID"
    ),
  ];
};
