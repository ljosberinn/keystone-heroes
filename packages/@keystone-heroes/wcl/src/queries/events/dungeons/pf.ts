import type {
  AllTrackedEventTypes,
  AnyEvent,
  ApplyBuffEvent,
  DamageEvent,
  DeathEvent,
} from "../types";
import {
  createChunkByThresholdReducer,
  createIsSpecificEvent,
  reduceEventsByPlayer,
} from "../utils";

const RED_BUFF = {
  unit: 164_705,
  aura: 340_225,
  // buff: 340_227,
} as const;

const PURPLE_BUFF = {
  unit: 164_707,
  aura: 340_271,
  // buff: 340_273,
} as const;

const GREEN_BUFF = {
  unit: 163_891,
  aura: 340_210,
  // buff: 340_211,
} as const;

export const PF = {
  PLAGUE_BOMB: 328_501,
  RIGGED_PLAGUEBORER: 168_878,
  GREEN_BUFF,
  RED_BUFF,
  PURPLE_BUFF,
} as const;

/**
 * @see https://www.warcraftlogs.com/reports/qGV26X4kLbRFBJDt#fight=5&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24(type%20%3D%20%22death%22%20and%20target.type%20%3D%20%22npc%22%20and%20target.id%20in%20(164705,164707,%20163891))%20or%20(type%20%3D%20%22applybuff%22%20and%20target.type%20%3D%20%22player%22%20and%20ability.id%20in%20(340225,%20340271,%20340210))%20or%20(type%20%3D%20%22damage%22%20and%20ability.id%20%3D%20328501)
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "qGV26X4kLbRFBJDt") {
 *       fights(fightIDs: [5]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 3958680, endTime: 6113126, filterExpression: "(type = \"death\" and target.type = \"npc\" and target.id in (164705,164707, 163891)) or (type = \"applybuff\" and target.type = \"player\" and ability.id in (340225, 340271, 340210)) or (type = \"damage\" and ability.id = 328501)") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "death" and target.type = "npc" and target.id in (${RED_BUFF.unit}, ${PURPLE_BUFF.unit}, ${GREEN_BUFF.unit})`,
  `type = "applybuff" and target.type = "player" and ability.id in (${RED_BUFF.aura}, ${PURPLE_BUFF.aura}, ${GREEN_BUFF.aura})`,
  `type = "damage" and ability.id = ${PF.PLAGUE_BOMB}`,
];

const createIsPfSlimeDeathEvent =
  (actorIDSet: Set<number>) =>
  (event: AnyEvent): event is DeathEvent => {
    return (
      event.type === "death" &&
      !actorIDSet.has(event.targetID) &&
      "targetInstance" in event &&
      event.targetInstance !== undefined
    );
  };

const isPfSlimeBuffEvent = createIsSpecificEvent<ApplyBuffEvent>({
  type: "applybuff",
  abilityGameID: [GREEN_BUFF.aura, RED_BUFF.aura, PURPLE_BUFF.aura],
});

const isPfPlagueBombDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: PF.PLAGUE_BOMB,
});

// ApplyBuff events fire a lot for the PF Slime Auras for some reason
// Fleshcraft gets its CD reduced, but its unlikely to reduce it by 75%
const pfSlimeBuffReducer = createChunkByThresholdReducer(30 * 1000);

export const getPFEvents = (
  allEvents: AllTrackedEventTypes[],
  actorIDSet: Set<number>
): (DeathEvent | DamageEvent | ApplyBuffEvent)[] => {
  const plagueBombDamageEvents = allEvents.filter(isPfPlagueBombDamageEvent);

  const plagueBombDamageTakenEvents = reduceEventsByPlayer(
    plagueBombDamageEvents.filter((event) => actorIDSet.has(event.targetID)),
    "targetID"
  );
  const plagueBombDamageDoneEvents = plagueBombDamageEvents.filter(
    (event) => !actorIDSet.has(event.targetID)
  );

  const aggregatedPlagueBombDamageEvent =
    plagueBombDamageDoneEvents.length > 0
      ? plagueBombDamageDoneEvents.reduce((acc, event) => {
          if (acc === event) {
            return acc;
          }

          return {
            ...acc,
            amount: acc.amount + event.amount - (event.overkill ?? 0),
          };
        }, plagueBombDamageDoneEvents[0])
      : null;

  const buffEvents = allEvents.filter(isPfSlimeBuffEvent).reduce<{
    red: ApplyBuffEvent[];
    green: ApplyBuffEvent[];
    purple: ApplyBuffEvent[];
  }>(
    (acc, event) => {
      switch (event.abilityGameID) {
        case GREEN_BUFF.aura:
          return { ...acc, green: [...acc.green, event] };
        case RED_BUFF.aura:
          return { ...acc, red: [...acc.red, event] };
        case PURPLE_BUFF.aura:
          return { ...acc, purple: [...acc.purple, event] };
      }

      return acc;
    },
    {
      red: [],
      green: [],
      purple: [],
    }
  );

  return [
    ...allEvents.filter(createIsPfSlimeDeathEvent(actorIDSet)),
    ...buffEvents.green
      .reduce<ApplyBuffEvent[][]>(pfSlimeBuffReducer, [])
      .flatMap((chunk) => chunk[0]),
    ...buffEvents.red
      .reduce<ApplyBuffEvent[][]>(pfSlimeBuffReducer, [])
      .flatMap((chunk) => chunk[0]),
    ...buffEvents.purple
      .reduce<ApplyBuffEvent[][]>(pfSlimeBuffReducer, [])
      .flatMap((chunk) => chunk[0]),
    ...plagueBombDamageTakenEvents,
    ...(aggregatedPlagueBombDamageEvent
      ? [aggregatedPlagueBombDamageEvent]
      : []),
  ];
};
