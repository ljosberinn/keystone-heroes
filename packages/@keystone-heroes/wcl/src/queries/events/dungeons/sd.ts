import type {
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyBuffStackEvent,
  BeginCastEvent,
  RemoveBuffEvent,
} from "../types";
import { createIsSpecificEvent } from "../utils";

export const SD_LANTERN_OPENING = 340_013;
export const SD_LANTERN_BUFF = 340_433;

/**
 * @see https://www.warcraftlogs.com/reports/mQn47kFCbpThBtrG#fight=15&type=summary&view=events&pins=2%24Off%24%23909049%24expression%24ability.id%20%3D%20340433%20and%20type%20in%20(%22applybuff%22,%20%22applybuffstack%22,%20%22removebuff%22)%5E2%24Off%24%23909049%24expression%24ability.id%20%3D%20340013%20and%20type%20%3D%22begincast%22
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "mQn47kFCbpThBtrG") {
 *       fights(fightIDs: [15]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 11536849, endTime: 13848195, filterExpression: "(type = \"begincast\" and ability.id = 340013) or (ability.id = 340433 and type in (\"applybuff\", \"applybuffstack\", \"removebuff\"))") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
const lanternOpeningExpression = `type = "begincast" and ability.id = ${SD_LANTERN_OPENING}`;
const lanternBuffExpression = `ability.id = ${SD_LANTERN_BUFF} and type in ("applybuff", "applybuffstack", "removebuff")`;

export const filterExpression = [
  lanternOpeningExpression,
  lanternBuffExpression,
];

const isSdLanternOpeningEvent = createIsSpecificEvent<BeginCastEvent>({
  type: "begincast",
  abilityGameID: SD_LANTERN_OPENING,
});

const isSdLanternBuffEvent = createIsSpecificEvent<ApplyBuffEvent>({
  type: "applybuff",
  abilityGameID: SD_LANTERN_BUFF,
});

const isSdLanternRemoveBuffEvent = createIsSpecificEvent<RemoveBuffEvent>({
  type: "removebuff",
  abilityGameID: SD_LANTERN_BUFF,
});

const isSdLanternApplyBuffStackEvent =
  createIsSpecificEvent<ApplyBuffStackEvent>({
    type: "applybuffstack",
    abilityGameID: SD_LANTERN_BUFF,
  });

export const getSDEvents = (
  allEvents: AllTrackedEventTypes[]
): (
  | BeginCastEvent
  | ApplyBuffEvent
  | RemoveBuffEvent
  | ApplyBuffStackEvent
)[] => {
  return allEvents.reduce<
    (BeginCastEvent | ApplyBuffEvent | RemoveBuffEvent | ApplyBuffStackEvent)[]
  >((acc, event) => {
    if (
      isSdLanternBuffEvent(event) ||
      isSdLanternOpeningEvent(event) ||
      isSdLanternRemoveBuffEvent(event) ||
      isSdLanternApplyBuffStackEvent(event)
    ) {
      return [...acc, event];
    }

    return acc;
  }, []);
};
