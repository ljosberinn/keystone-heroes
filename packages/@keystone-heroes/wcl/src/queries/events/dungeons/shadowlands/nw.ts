import type { DamageEvent, HealEvent } from "../../types";
import {
  createChunkByThresholdReducer,
  createIsSpecificEvent,
} from "../../utils";

export const NW = {
  ORB: 328_406,
  HAMMER: 328_128,
  SPEAR: 328_351,
  KYRIAN_ORB: {
    heal: 344_422,
    damage: 344_421,
  },
};

/**
 * @see https://www.warcraftlogs.com/reports/Jq7KrbYV1hmTWMyw#fight=4&type=summary&view=events&pins=2%24Off%24%23909049%24expression%24type%20%3D%20%22heal%22%20and%20source.type%20%3D%20%22player%22%20and%20ability.id%20%3D344422%5E2%24Off%24%23909049%24expression%24type%20%3D%20%22damage%22%20and%20source.type%20%3D%20%22player%22%20and%20ability.id%20%3D328406%5E2%24Off%24%23a04D8A%24expression%24type%20%3D%20%22damage%22%20and%20source.type%20%3D%20%22player%22%20and%20ability.id%20%3D328128%5E2%24Off%24%23DF5353%24expression%24type%20%3D%20%22damage%22%20and%20source.type%20%3D%20%22player%22%20and%20ability.id%20%3D328351%5E2%24Off%24rgb(78%25,%2061%25,%2043%25)%24expression%24type%20%3D%20%22damage%22%20and%20source.type%20%3D%20%22player%22%20and%20ability.id%20%3D344421
 */
const orbDamageFilterExpression = `type = "damage" and source.type = "player" and ability.id = ${NW.ORB}`;
const hammerDamageFilterExpression = `type = "damage" and source.type = "player" and ability.id = ${NW.HAMMER}`;
const spearDamageFilterExpression = `type = "damage" and source.type = "player" and ability.id = ${NW.SPEAR}`;
const kyrianOrbDamageFilterExpression = `type = "damage" and source.type = "player" and ability.id = ${NW.KYRIAN_ORB.damage}`;
const kyrianHealFilterExpression = `type = "heal" and source.type = "player" and ability.id = ${NW.KYRIAN_ORB.heal}`;

export const filterExpression = [
  orbDamageFilterExpression,
  hammerDamageFilterExpression,
  spearDamageFilterExpression,
  kyrianHealFilterExpression,
  kyrianOrbDamageFilterExpression,
];

export const isNwSpearEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: NW.SPEAR,
});

export const isNwHammerEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: NW.HAMMER,
});

export const isNwOrbEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: NW.ORB,
});

export const isNwKyrianOrbDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: NW.KYRIAN_ORB.damage,
});

export const isNwKyrianOrbHealEvent = createIsSpecificEvent<HealEvent>({
  type: "heal",
  abilityGameID: NW.KYRIAN_ORB.heal,
});

// NW Spear applies a bleed for 16 seconds
// each usage is hopefully thus at least 16s apart of each other
// may have to adjust later for multi-spearing...
export const nwSpearReducer = createChunkByThresholdReducer(16 * 1000);

// NW Orb pulses every 1s for 8s
// each usage is hopefully thus at least 8s apart of each other
// may have to adjust later for multi-orbing...
export const nwOrbReducer = createChunkByThresholdReducer(8 * 1000);
