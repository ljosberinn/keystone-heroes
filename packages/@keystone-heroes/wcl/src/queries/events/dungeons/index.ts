import { DungeonIDs } from "@keystone-heroes/db/data";
import type { PlayableClass } from "@keystone-heroes/db/types";

import type {
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyBuffStackEvent,
  ApplyDebuffEvent,
  BeginCastEvent,
  CastEvent,
  DamageEvent,
  DeathEvent,
  HealEvent,
  RemoveBuffEvent,
} from "../../events/types";
import { filterExpression as dosFilterExpression, getDOSEvents } from "./dos";
import { filterExpression as hoaFilterExpression, getHOAEvents } from "./hoa";
import { filterExpression as nwFilterExpression, getNWEvents } from "./nw";
import { filterExpression as pfFilterExpression, getPFEvents } from "./pf";
import { filterExpression as sdFilterExpression, getSDEvents } from "./sd";
import { filterExpression as soaFilterExpression, getSOAEvents } from "./soa";
import { filterExpression as topFilterExpression, getTOPEvents } from "./top";

type DungeonWithEvents = Exclude<DungeonIDs, DungeonIDs.MISTS_OF_TIRNA_SCITHE>;

const isDungeonWithEvent = (id: DungeonIDs): id is DungeonWithEvents =>
  id in dungeonExpressionMap;

const dungeonExpressionMap: Record<DungeonWithEvents, string[]> = {
  [DungeonIDs.DE_OTHER_SIDE]: dosFilterExpression,
  [DungeonIDs.HALLS_OF_ATONEMENT]: hoaFilterExpression,
  [DungeonIDs.PLAGUEFALL]: pfFilterExpression,
  [DungeonIDs.SANGUINE_DEPTHS]: sdFilterExpression,
  [DungeonIDs.SPIRES_OF_ASCENSION]: soaFilterExpression,
  [DungeonIDs.THEATRE_OF_PAIN]: topFilterExpression,
  [DungeonIDs.THE_NECROTIC_WAKE]: nwFilterExpression,
};

export const getDungeonExpression = (id: DungeonIDs): string[] =>
  isDungeonWithEvent(id) ? dungeonExpressionMap[id] : [];

export const filterDungeonEvents = (
  allEvents: AllTrackedEventTypes,
  dungeonID: DungeonIDs,
  playerMetaInformation: { actorID: number; class: PlayableClass }[]
):
  | ApplyDebuffEvent[]
  | CastEvent[]
  | (DeathEvent | DamageEvent | ApplyBuffEvent)[]
  | (BeginCastEvent | ApplyBuffEvent | RemoveBuffEvent | ApplyBuffStackEvent)[]
  | (DamageEvent | HealEvent)[] => {
  switch (dungeonID) {
    case DungeonIDs.DE_OTHER_SIDE:
      return getDOSEvents(allEvents);
    case DungeonIDs.HALLS_OF_ATONEMENT:
      return getHOAEvents(allEvents);
    case DungeonIDs.PLAGUEFALL:
      return getPFEvents(allEvents, playerMetaInformation);
    case DungeonIDs.SANGUINE_DEPTHS:
      return getSDEvents(allEvents);
    case DungeonIDs.SPIRES_OF_ASCENSION:
      return getSOAEvents(allEvents);
    case DungeonIDs.THEATRE_OF_PAIN:
      return getTOPEvents(allEvents);
    case DungeonIDs.THE_NECROTIC_WAKE:
      return getNWEvents(allEvents);
    default:
      return [];
  }
};
