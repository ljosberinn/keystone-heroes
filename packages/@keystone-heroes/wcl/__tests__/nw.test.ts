import { getNWEvents } from "@keystone-heroes/wcl/queries/events/dungeons/nw";

import allEvents from "./fixtures/allEvents.json";

describe("getNWEvents", () => {
  test("works", () => {
    expect(getNWEvents(allEvents)).toMatchSnapshot();
  });
});
