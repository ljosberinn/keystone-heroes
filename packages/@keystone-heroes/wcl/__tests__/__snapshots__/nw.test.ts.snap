// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getNWEvents works 1`] = `
Array [
  Object {
    "abilityGameID": 344421,
    "amount": 98205,
    "hitType": 1,
    "pin": "0",
    "sourceID": 1,
    "sourceMarker": 2,
    "targetID": 10,
    "targetInstance": 10,
    "timestamp": 1913816,
    "type": "damage",
    "unmitigatedAmount": 4922,
  },
  Object {
    "abilityGameID": 344421,
    "amount": 102873,
    "hitType": 2,
    "pin": "0",
    "sourceID": 9,
    "targetID": 10,
    "targetInstance": 10,
    "timestamp": 1915015,
    "type": "damage",
    "unmitigatedAmount": 5282,
  },
  Object {
    "abilityGameID": 344421,
    "amount": 106175,
    "hitType": 1,
    "overkill": 229,
    "pin": "0",
    "sourceID": 4,
    "sourceMarker": 7,
    "targetID": 10,
    "targetInstance": 8,
    "timestamp": 1916508,
    "type": "damage",
    "unmitigatedAmount": 5333,
  },
  Object {
    "abilityGameID": 344421,
    "amount": 60626,
    "hitType": 1,
    "pin": "0",
    "sourceID": 8,
    "targetID": 10,
    "targetInstance": 10,
    "timestamp": 1927087,
    "type": "damage",
    "unmitigatedAmount": 4728,
  },
  Object {
    "abilityGameID": 344421,
    "amount": 73018,
    "hitType": 2,
    "pin": "0",
    "sourceID": 2,
    "targetID": 10,
    "targetInstance": 10,
    "timestamp": 1928443,
    "type": "damage",
    "unmitigatedAmount": 4821,
  },
  Object {
    "abilityGameID": 344422,
    "amount": 54149,
    "hitType": 1,
    "overheal": 95456,
    "pin": "0",
    "sourceID": 1,
    "sourceMarker": 2,
    "targetID": 1,
    "targetMarker": 2,
    "timestamp": 1913800,
    "type": "heal",
  },
  Object {
    "abilityGameID": 344422,
    "amount": 63747,
    "hitType": 2,
    "overheal": 131191,
    "pin": "0",
    "sourceID": 9,
    "targetID": 9,
    "timestamp": 1915001,
    "type": "heal",
  },
  Object {
    "abilityGameID": 344422,
    "amount": 47270,
    "hitType": 1,
    "overheal": 107811,
    "pin": "0",
    "sourceID": 4,
    "sourceMarker": 7,
    "targetID": 4,
    "targetMarker": 7,
    "timestamp": 1916507,
    "type": "heal",
  },
  Object {
    "abilityGameID": 344422,
    "amount": 5902,
    "hitType": 1,
    "overheal": 114954,
    "pin": "0",
    "sourceID": 8,
    "targetID": 8,
    "timestamp": 1923089,
    "type": "heal",
  },
  Object {
    "abilityGameID": 344422,
    "amount": 53236,
    "hitType": 1,
    "overheal": 89835,
    "pin": "0",
    "sourceID": 2,
    "targetID": 2,
    "timestamp": 1924449,
    "type": "heal",
  },
]
`;
